### PR TITLE
Contains MiniWallet Contracts refactor, MiniServer, Proxy Upgrade Deploys and NFT Scaffolding

### PR DESCRIPTION
**Pull request by johnwhitton**
This branch is being used to work on Minting NFT for users when registering  https://github.com/polymorpher/sms-wallet/issues/10

### CODE REVIEW

File: miniwallet/contracts/nft/HRC721/HRC721.sol [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+ * @dev Implementation of https://eips.ethereum.org/EIPS/eip-721[ERC721] Non-Fungible Token Standard, including
+ * the Metadata and Enumerable extension. Built to optimize for lower gas during batch mints.
+ *
+ * Assumes serials are sequentially minted starting at 0 (e.g. 0, 1, 2, 3..).
+ *
+ * Assumes that an owner cannot have more than 2**64 - 1 (max value of uint64) of supply.
+ *
+ * Assumes that the maximum token id cannot exceed 2**128 - 1 (max value of uint128).
+ */
+contract HRC721 is
```
Comment (polymorpher) [inline miniwallet/contracts/nft/HRC721/HRC721.sol:45]: Let's use a more standard name (ERC721 followed by something)

File: miniwallet/test/utilities/Mini721Util.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  const mintPrice = await thisObject.mini721.mintPrice()
+  // console.log(`mintPrice: ${JSON.stringify(mintPrice)}`)
+  const tx = await thisObject.mini721.connect(minter).mintMini(numTokens, {
+    value: mintPrice.mul(numTokens)
+  })
+  return tx
+}
+
+export async function communityMint721 (thisObject, owner, numTokens) {
+  const tx = await thisObject.mini721.mintForCommunity(owner, numTokens)
```
Comment (polymorpher) [inline miniwallet/test/utilities/Mini721Util.ts:50]: could simplify to one line. `thisObject` should also be replaced with something more informative

File: miniwallet/test/utilities/Mini721Util.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'
+  thisObject.bobPrivateKey =
+    '0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d'
+  thisObject.carolPrivateKey =
+    '0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a'
+  thisObject.Mini721 = await ethers.getContractFactory('Mini721')
+}
+
+export async function deploy721 (thisObject, contracts) {
+  for (const i in contracts) {
```
Comment (polymorpher) [inline miniwallet/test/utilities/Mini721Util.ts:33]: `for (const contract of contracts) {`

File: miniwallet/test/utilities/Mini1155Util.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
+// import { expect } from 'chai'
+import { ethers, network } from 'hardhat'
+
+// import { userMint1155, safeTransferFrom1155 } from './utilities'
+
+export async function mini1155configure ({
+  mini1155,
+  config = {
```
Comment (polymorpher) [inline miniwallet/test/utilities/Mini1155Util.ts:8]: let's use more general config

File: miniwallet/test/utilities/Mini1155Util.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  numTokens
+) {
+  const tx = await mini1155
+    .connect(sender)
+    .safeTransferFrom(
+      sender.address,
+      receiver.address,
+      tokenId,
+      numTokens,
+      '0x'
```
Comment (polymorpher) [inline miniwallet/test/utilities/Mini1155Util.ts:70]: ''

File: miniwallet/test/utilities/Mini1155Util.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+export async function mini1155Mint (mini1155, minter, numTokens) {
+  const tx = await mini1155
+    .connect(minter)['mint(uint256)'](numTokens)
+  return tx
+}
+
+export async function mini1155OwnerMint (mini1155, minter, tokenId, numTokens) {
+  const tx = await mini1155.connect(minter)['mint(address,uint256,uint256,bytes)'](minter, tokenId, numTokens, '0x')
```
Comment (polymorpher) [inline miniwallet/test/utilities/Mini1155Util.ts:52]: ''

File: miniwallet/test/Mini721.ts [hunk 7]
```diff
@@ -0,0 +1,971 @@
...
+    // console.log(`Reverting Snapshot : ${snapshotId}`);
+    await network.provider.send('evm_revert', [this.snapshotId])
+  })
+  // TODO
+  // NFT Build Process (Images, Metadata, Provenance Hash)
+
+  // Deployment Tests
+  describe('validate Initial Deploy', function (this) {
+    it('Mini721 contract should be created', async function () {
+      expect(await this.mini721.address).to.equal(
```
Comment (polymorpher) [inline miniwallet/test/Mini721.ts:40]: this might be too specific

File: miniwallet/contracts/nft/Mini721.sol [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+    function setBaseUri(string memory _baseUri)
+        external
+        onlyOwner
+        whenMetadataNotFrozen
+    {
+        baseUri = _baseUri;
+        emit SetBaseUri(baseUri);
+    }
+
+    function mintForCommunity(address _to, uint256 _numberOfTokens)
```
Comment (polymorpher) [inline miniwallet/contracts/nft/Mini721.sol:283]: Let's have a batch version of this, or have a separate multi-call contract. I recall it wasn't a nice experience using this for thousands of airdrops

File: miniwallet/test/Mini721.ts [hunk 9]
```diff
@@ -0,0 +1,971 @@
...
+      expect(await this.mini721.temporaryTokenUri()).to.equal('')
+
+      // Public Functions Mini721
+      await expect(this.mini721.tokenURI(0)).to.be.revertedWith(
+        'URIQueryForNonexistentToken()'
+      )
+      expect(await this.mini721.supportsInterface('0x2a55205a')).to.equal(true)
+      expect(await this.mini721.exists(18)).to.equal(false)
+      expect(await this.mini721.contractURI()).to.equal('')
+      // Token should hold no eth initially
```
Comment (polymorpher) [inline miniwallet/test/Mini721.ts:66]: remove some commented out code?

File: miniwallet/test/Mini721.ts [hunk 10]
```diff
@@ -0,0 +1,971 @@
...
+        'ipfs://TemporaryTokenUriPlaceHolder/'
+      )
+
+      // Public Functions Mini721
+      expect(await this.mini721.tokenURI(12)).to.equal(
+        'ipfs://TemporaryTokenUriPlaceHolder/'
+      )
+
+      // TODO review supportsInterface logic
+      expect(await this.mini721.supportsInterface('0x2a55205a')).to.equal(true)
```
Comment (polymorpher) [inline miniwallet/test/Mini721.ts:287]: should have a constant mapping somewhere to human-readable methods

File: miniwallet/test/Mini721.ts [hunk 11]
```diff
@@ -0,0 +1,971 @@
...
+      //   "Mini721: Metadata is frozen"
+      // );
+      // await expect(this.mini721.batchBurn([10, 11, 13])).to.be.revertedWith(
+      //   "Mini721: Metadata is frozen"
+      // );
+    })
+  })
+
+  // NFT Royalty Testing
+  describe('Mini721Royalties', function (this) {
```
Comment (polymorpher) [inline miniwallet/test/Mini721.ts:877]: not needed

File: miniwallet/test/Mini721.ts [hunk 12]
```diff
@@ -0,0 +1,971 @@
...
+  describe('Mini721TestEvents', function (this) {
+    it('Mini721 Test all events are emitted correctly', async function () {
+      // safeTransferFrom(this.carol.address, this.bob.address, 13)
+      // safeTransferFrom(this.carol.address, this.bob.address, 13, "bytes")
+      // getRaribleV2Royalties(uint256)
+    })
+  })
+
+  // Test remaining functions including royalties and granitng approval
+  describe('MiniRemainingFunctions', function (this) {
```
Comment (polymorpher) [inline miniwallet/test/Mini721.ts:895]: probably not needed

File: miniwallet/test/Mini721.ts [hunk 13]
```diff
@@ -0,0 +1,971 @@
...
+  describe('Mini721Royalties', function (this) {
+    it('Mini721 Test Royalties', async function () {
+      // safeTransferFrom(this.carol.address, this.bob.address, 13)
+      // safeTransferFrom(this.carol.address, this.bob.address, 13, "bytes")
+      // getRaribleV2Royalties(uint256)
+    })
+  })
+
+  // NFT Event Testing
+  describe('Mini721TestEvents', function (this) {
```
Comment (polymorpher) [inline miniwallet/test/Mini721.ts:886]: let's leave it for later

File: miniwallet/test/Mini721.ts [hunk 14]
```diff
@@ -0,0 +1,971 @@
...
+      expect(await this.mini721.ownerOf(2201)).to.equal(this.carol.address)
+      await this.mini721
+        .connect(this.carol)
+        .transferFrom(this.carol.address, this.alice.address, 2201)
+      expect(await this.mini721.ownerOf(1)).to.equal(this.alice.address)
+    })
+  })
+  // 7. Metadata Update (fixing outdated metadata)
+  // MetaData
+  describe('checkMetaData', function (this) {
```
Comment (polymorpher) [inline miniwallet/test/Mini721.ts:954]: already tested above

File: miniwallet/test/Mini721.ts [hunk 15]
```diff
@@ -0,0 +1,971 @@
...
+      // setTemporaryTokenUri(string)
+      // setStartIndex()
+      // setUri(uint256,string)
+
+    })
+  })
+
+  // NFT (Transfer Ownership)
+  describe('Mini721NFTSale', function (this) {
+    it('Mini721 tokens should be sold by users', async function () {
```
Comment (polymorpher) [inline miniwallet/test/Mini721.ts:922]: seems to be just testing transfer

File: miniwallet/test/Mini1155.ts [hunk 16]
```diff
@@ -0,0 +1,597 @@
+import { expect } from 'chai'
+import { ethers, network } from 'hardhat'
+
+import { mini1155Mint, mini1155SafeTransferFrom, mini1155configure } from './utilities'
+
+export const ADDRESS_ZERO = '0x0000000000000000000000000000000000000000'
+
+// Deployment configuration
+const d = {
```
Comment (polymorpher) [inline miniwallet/test/Mini1155.ts:9]: let's use more general parameters

File: miniwallet/test/Mini1155.ts [hunk 17]
```diff
@@ -0,0 +1,597 @@
...
+
+      // Pause Contract (Owner)
+      await this.mini1155.pause()
+      await this.mini1155.unpause()
+
+      // Freeze Metadata (Owner)
+      await this.mini1155.freezeMetadata()
+    })
+
+    it('Mini1155-9 Royalty Validation', async function () {
```
Comment (polymorpher) [inline miniwallet/test/Mini1155.ts:589]: skip?

File: miniwallet/test/Mini1155.ts [hunk 18]
```diff
@@ -0,0 +1,597 @@
...
+      // Do multiple transfers and check the royalties account balance gets updated.
+    })
+
+    it('Mini1155-10 GAS Usage', async function () {
```
Comment (polymorpher) [inline miniwallet/test/Mini1155.ts:593]: probably not needed

**Review by polymorpher — COMMENTED** [review]
overall looks good aside from minor cosmetic and syntactical changes

File: miniwallet/contracts/nft/HRC721/HRC721.sol (hunk 1 above)
Comment (johnwhitton) [reply miniwallet/contracts/nft/HRC721/HRC721.sol:45]: Done using ERC721M (for MiniWallet)

File: miniwallet/test/utilities/Mini721Util.ts (hunk 2 above)
Comment (johnwhitton) [reply miniwallet/test/utilities/Mini721Util.ts:50]: done 🙏

File: miniwallet/test/utilities/Mini1155Util.ts (hunk 4 above)
Comment (johnwhitton) [reply miniwallet/test/utilities/Mini1155Util.ts:8]: done 🙏

File: miniwallet/test/utilities/Mini1155Util.ts (hunk 5 above)
Comment (johnwhitton) [reply miniwallet/test/utilities/Mini1155Util.ts:70]: done 🙏

File: miniwallet/test/Mini721.ts (hunk 7 above)
Comment (johnwhitton) [reply miniwallet/test/Mini721.ts:40]: Changed to `.to.be.properAddress;` 
https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#proper-address
🙏

File: miniwallet/test/Mini721.ts (hunk 9 above)
Comment (johnwhitton) [reply miniwallet/test/Mini721.ts:66]: cleaned up 🙏

File: miniwallet/test/utilities/Mini721Util.ts (hunk 3 above)
Comment (johnwhitton) [reply miniwallet/test/utilities/Mini721Util.ts:33]: done 🙏

File: miniwallet/test/utilities/Mini1155Util.ts (hunk 6 above)
Comment (johnwhitton) [reply miniwallet/test/utilities/Mini1155Util.ts:52]: done 🙏

File: miniwallet/test/Mini721.ts (hunk 10 above)
Comment (johnwhitton) [reply miniwallet/test/Mini721.ts:287]: Added this in constants 🙏

File: miniwallet/test/Mini721.ts (hunk 12 above)
Comment (johnwhitton) [reply miniwallet/test/Mini721.ts:895]: removed for now 🙏

File: miniwallet/test/Mini721.ts (hunk 14 above)
Comment (johnwhitton) [reply miniwallet/test/Mini721.ts:954]: removed 🙏

File: miniwallet/test/Mini721.ts (hunk 15 above)
Comment (johnwhitton) [reply miniwallet/test/Mini721.ts:922]: Updated description 🙏

File: miniwallet/test/Mini1155.ts (hunk 17 above)
Comment (johnwhitton) [reply miniwallet/test/Mini1155.ts:589]: removed for now 🙏

File: miniwallet/test/Mini1155.ts (hunk 18 above)
Comment (johnwhitton) [reply miniwallet/test/Mini1155.ts:593]: removed for now 🙏

File: miniwallet/test/Mini721.ts (hunk 13 above)
Comment (johnwhitton) [reply miniwallet/test/Mini721.ts:886]: removed for now 🙏

File: miniwallet/test/Mini1155.ts (hunk 16 above)
Comment (johnwhitton) [reply miniwallet/test/Mini1155.ts:9]: done 🙏

File: miniwallet/test/Mini721.ts (hunk 11 above)
Comment (johnwhitton) [reply miniwallet/test/Mini721.ts:877]: removed for now 🙏

**Comment by johnwhitton** [comment]
Outstanding Items
- [ ] Multi-call MintForCommunity
- [ ] Deployment Scripts
- [ ] Update NFT UI to use deployed contracts as DUMMY_TOKENS
- [ ] Registration WorkFlow

**Comment by johnwhitton** [comment]
Closing this PR will progress under PR #14 
Items are being tracked under #13

**Pull request closed without merging**